Updating .NET MAUI to use MauiApp/MauiAppBuilder pattern and use MS.Extensions.DependencyInjection

Eilon opened the pull request:
### What's new/different:

- Each platform-specific application type is abstract and has an abstract method `protected abstract MauiAppBuilder CreateAppBuilder();` that is overridden in the user's app to create and return a `MauiAppBuilder`.
- The `IStartup` interface is removed, and the user's app startup is in a conventional `MauiProgram` class with a static method that creates the app builder and returns the built `MauiApp`.
- `MauiAppBuilder` is akin to ASP.NET Core's [WebApplicationBuilder](https://github.com/dotnet/aspnetcore/blob/main/src/DefaultBuilder/src/WebApplicationBuilder.cs) and has properties, methods, and extension methods for setting up the app. For example, registering services, fonts, handlers, etc.
- All of Microsoft.Extensions.DependencyInjection's features are supported (constructor injection, enumerable multi-services, open generic registrations, etc.).

Before:

```c#
	// Platform-independent app setup
	public class Startup : IStartup
	{
		public void Configure(IAppHostBuilder appBuilder)
		{
			appBuilder
				.UseMauiApp<App>()
				.ConfigureFonts(fonts =>
				{
					fonts.AddFont("OpenSans-Regular.ttf", "OpenSansRegular");
				});
		}
	}

	// Each platform (WinUI/Android/iOS/MacCatalyst) has something like this:
	public class MainApplication : MauiApplication<Startup>
	{
		...
	}
```

After:

```c#
	// Platform-independent app setup
	public static class MauiProgram
	{
		public static MauiApp CreateMauiApp()
		{
			var builder = MauiApp.CreateBuilder();
			builder
				.UseMauiApp<App>()
				.ConfigureFonts(fonts =>
				{
					fonts.AddFont("OpenSans-Regular.ttf", "OpenSansRegular");
				});
			return builder.Build();
		}
	}

	// Each platform (WinUI/Android/iOS/MacCatalyst) has something like this:
	public class MainApplication : MauiApplication
	{
		...

		protected override MauiApp CreateMauiApp() => MauiProgram.CreateMauiApp();
	}
```

Please share feedback in this PR, either in general, or on specific lines of code.

Eilon committed:
WIP: Startup/host unification

davidfowl commented on src/BlazorWebView/src/Maui/BlazorWebViewRegistrationExtensions.cs:
```diff
@@ -1,18 +1,19 @@
 	public static class BlazorWebViewRegistrationExtensions
 	{
-		public static TAppHostBuilder RegisterBlazorMauiWebView<TAppHostBuilder>(this TAppHostBuilder appHostBuilder) where TAppHostBuilder : IAppHostBuilder
+		public static MauiAppBuilder RegisterBlazorMauiWebView(this MauiAppBuilder appHostBuilder)
```
Does it need to be a callback?

```
appHostBuilder.RegisterBlazorMauiWebView()
```

davidfowl commented on src/Core/src/Core/MauiApp.cs:
```diff
@@ -0,0 +1,87 @@
+
+namespace Microsoft.Maui
+{
+	public sealed class MauiApp : IHost
```
MaulApplication?

```suggestion
	public sealed class MauiApplication : IHost
```

Eilon replied:
Unfortunately there are already many various types in MAUI called MauiApplication or similar, which contain the user's application entry point on each platform. I don't like the name `MauiApp` so for now it's a placeholder until we can figure out the right names for everything.

davidfowl replied:
I see.

mattleibow replied:
Since it is IHost, we can use that word. It is not the app, so MauiAppHost or MauiHost?

PureWeen commented on src/Compatibility/ControlGallery/src/Core/Startup.cs:
```diff
@@ -26,14 +27,15 @@ public virtual void Configure(IAppHostBuilder appBuilder)
 				});
+
+			// TODO: This used to be called inside a callback from ConfigureServices, but doesn't actually use any services. Was it just to delay execution to be later in app startup?
+			DependencyService.Register(Device.GetAssemblies());
```
It's fine if you call it from here

Eilon commented:
All the stuff I said was _un-done_ is now _done_. Support for IHost, IHostBuilder, custom DI, config is all done (by copying stuff from AspNetCore). I _believe_ that all tests pass because they did pass locally not long ago, and CI seems _fairly_ happy, aside from things that are flaky even in main. There is likely still _some_ bug in the code anyway, which I will continue to hunt down.

What's left now?

* Finalize naming of major public types, such as `MauiAppBuilder` (though this type _name_ is rarely seen in user code), and `MauiApp` (which is typically seen at least once in projects to create the builder).
* Lots of _code cleanup_ to do (rename files/types, etc.), but I would much rather wait until the rest is committed before getting to that because I tried to keep things in such a way that I could more easily keep re-basing on main. (If I move things _too_ much, it's nearly impossible to rebase.)
* Add any _new_ tests that are needed. (Almost all existing tests continue to exist and were merely updated; a few that are not relevant anymore were deleted.)
* Project template updates (this is trivial, and I'm also delaying it to avoid merge conflicts_

@redth / @PureWeen / @rmarinho / @mattleibow / @davidfowl - I would love your feedback on what we've got here so that we can get this into RC1.

Eilon commented on eng/Microsoft.Extensions.targets:
```diff
@@ -3,19 +3,7 @@
     <_UseNet6Packages Condition=" '$(TargetFramework.Contains(net6.0))' == 'true' or '$(PackageType)' == 'DotnetPlatform' ">true</_UseNet6Packages>
   </PropertyGroup>
   <PropertyGroup Condition=" '$(_UseNet6Packages)' != 'true' ">
-    <_MicrosoftHostingVersion>5.0.0</_MicrosoftHostingVersion>
```
I dropped all these versions because they don't seem to be needed. All the 6.0.0 versions of these packages continue to support the same TFMs as before, so I don't think there's any need to restrict parts of the build to the 5.0.0 versions. The _real_ reason I removed it is to avoid some build issue where diamond dependencies were causing NuGet errors. Either way, not needed as far as I know.

Eilon commented:
Local test pass in WinUI SLN (everything passes except some tests that aren't supposed to pass locally, and I don't know how to run DeviceTests).

![image](https://user-images.githubusercontent.com/202643/130300723-5e72a90f-43ab-4075-a028-b4b19d58686b.png)

Eilon commented:
And for fun I just tried using Autofac as the DI container and it seems to work great when I run in Android! I referred to the [Autofac docs for ASP.NET Core](https://autofac.readthedocs.io/en/latest/integration/aspnetcore.html) because that's the pattern in this PR for .NET MAUI.

![image](https://user-images.githubusercontent.com/202643/130395154-7c13b64c-5fd5-43f7-af23-9f7ffe1554a1.png)

I added a reference to this Autofac package:

```xml
    <PackageReference Include="Autofac.Extensions.DependencyInjection" Version="7.1.0" />
```

And here's what I changed in Startup.cs to make use of Autofac:

```c#
using Autofac;
using Autofac.Extensions.DependencyInjection;

...

	public static MauiAppBuilder CreateAppBuilder()
	{
		var appBuilder = MauiApp.CreateBuilder();

		appBuilder.Host.UseServiceProviderFactory(new AutofacServiceProviderFactory(containerBuilder =>
		{
			// Before (with Microsoft.Extensions.DependencyInjection registration):
			//services.AddSingleton<ITextService, TextService>();
			//services.AddTransient<MainViewModel>();

			// After (with Autofac registration):
			containerBuilder.RegisterType<TextService>().As<ITextService>().SingleInstance();
			containerBuilder.RegisterType<MainViewModel>().AsSelf().InstancePerDependency();
		}));
		...
	}
```

And all the rest of the registrations in the app were done using MS.Ext.DI registration because it's _mostly_ various .NET MAUI and Blazor stuff.

PureWeen commented on src/Core/src/Core/FontsMauiAppBuilderExtensions.cs:
```diff
@@ -0,0 +1,82 @@
+{
+	public static class FontsMauiAppBuilderExtensions
+	{
+		public static MauiAppBuilder ConfigureFonts(this MauiAppBuilder builder)
```
Is the plan to eventually collapse this (and similar ones) down as well?

either to

```C#
appBuilder.Fonts.AddFont
```

or
```C#
appBuilder.AddFont
```

mattleibow replied:
I was on the fence about this for no reason than we need some sort of process for identifying what we want to add there. Once it is added, it can't really be removed. Maybe fonts is not bad, but we need a process of ensuring we don't willy-nilly add properties. We can also help people by renaming this type `Forms` 😉 

It is also breaking to change the property type.

Fonts might be good to add, but we need to make sure we got rules!

davidfowl replied:
I like appBuilder.Fonts.*

mattleibow replied:
How does ASP.NET handle new things? Is it just added to the type as a property? How do you prevent growth for everything? 

We have app lifecycle events, platform effects (which is maybe something that is old and should not be used), fonts, image source handlers, and possibly a few others. We will end up with many things on there. 

I suppose it is not _bad_, but may cause issues later on if we got things on there that have changed and now no longer make sense:

```csharp
class MauiAppBuilder {
    IServiceCollection Services;
    ConfigurationManager Configuration;
    ILoggingBuilder Logging;
    ConfigureHostBuilder Host;
    IFontCollection Fonts;
    IEffectsCollection Effects;
    ILifecycleEventManager LifecycleEvents;
    IImageSources ImageSources;
}
```

davidfowl replied:
There are 2 things being created here:
- The builder (wrapper around the host)
- The application itself (this is where you'll see most ASP.NET primitives exposed). These are things in the service provider that 90% of application will try to access top level https://github.com/dotnet/aspnetcore/blob/3c089fa29a441a390f2786e5107f0dcc8397898f/src/DefaultBuilder/src/WebApplication.cs#L40-L65.

The builder exposes `IHostBuilder` and `IWebHostBuilder` property because the existing ecosystem need to be able to configure more advanced things.

Generally, I think you want to use the 80/20 rule while determining what to expose top level.

mattleibow replied:
@PureWeen  then maybe Fonts is cool here. I think things like effects and image sources are more for library authors. Fonts and maybe lifecycle events are user app things. Although, lifecycle events are a bit rare I think? Since most people using that will either be people doing native libraries or library authors.

Maybe just fonts?

mattleibow replied:
And handlers. I forgot handlers are like essential for anything.

mattleibow commented on src/Core/src/Core/HandlerMauiAppBuilderExtensions.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Microsoft.Maui
+{
+	public static class HandlerMauiAppBuilderExtensions
```
I am thinking that since handlers are a core function of Maui, this should be a property on the AppBuilder? I know I just made a comment saying we need rules, but I think this is so fundamental that an app can't run without it.

mattleibow commented on src/Core/src/Hosting/LifecycleEvents/AppHostBuilderExtensions.cs:
```diff
@@ -1,36 +1,34 @@
+		public static MauiAppBuilder ConfigureLifecycleEvents(this MauiAppBuilder builder, Action<ILifecycleBuilder>? configureDelegate)
 		{
-			public void ConfigureServices(HostBuilderContext context, IServiceCollection services)
+			builder.Services.AddSingleton<ILifecycleEventService, LifecycleEventService>();
```
One thing that I am noticing with the new way is that this service is getting registered multiple times. Is that going to cause issues?

```csharp
builder.ConfigureLifecycleEvents(app => app.Xxx());
...
builder.ConfigureLifecycleEvents(app => app.Xxx());
```

Each call to `ConfigureLifecycleEvents` will re-register `LifecycleEventService`

Eilon replied:
Yeah most (or all?) of the framework calls I need to change to TRY AddXyz and that will fix this.

mattleibow commented on src/Core/src/Core/MauiApp.cs:
```diff
@@ -0,0 +1,87 @@
+		/// A <see cref="Task"/> that represents the startup of the <see cref="MauiApp"/>.
+		/// Successful completion indicates the HTTP server is ready to accept new requests.
+		/// </returns>
+		public Task StartAsync(CancellationToken cancellationToken = default) =>
```
Start and stop should probably be implemented explicitly because I don't think they are used in maui

Eilon replied:
I think it could be nice to fully support IApplicationLifetime et al so I'll look into what we can do here.

mattleibow replied:
Yeah, _we_ may not use it for maui, but we still want Bob's library to work.

mattleibow commented:
@Eilon you can run device test by just launching those apps. It is just a normal maui app with a UI and everything. There is a "Run All" button.

Eilon commented:
I just had a great meeting with @halter73 from the ASP.NET Core team to review some of the patterns here.

These are my notes to investigate and implement from the feedback in this PR from various folks and also from my meeting with @halter73 .

- [ ] Consider naming of the .NET MAUI per-platform application base types in terms of consistency across platforms and in terms of using up the "good" names for types that are almost never seen. For example, on WinUI there's [`MauiWinUIApplication`](https://github.com/dotnet/maui/blob/main/src/Core/src/Platform/Windows/MauiWinUIApplication.cs), on Android there's [`MauiApplication`](https://github.com/dotnet/maui/blob/main/src/Core/src/Platform/Android/MauiApplication.cs), on iOS/MacCatalyst there's [`MauiUIApplicationDelegate`](https://github.com/dotnet/maui/blob/main/src/Core/src/Platform/iOS/MauiUIApplicationDelegate.cs). Perhaps these should either use names more similar to each other (not just `Maui` prefixed to the base class), or at least not use up very generic names like `MauiApplication` that we could instead use somewhere else.
- [ ] Consider using the "good" application types names for the more user-seen parts of the app, and change `MauiApp`-->`MauiApplication` and `MauiAppBuilder`-->`MauiApplicationBuilder`
- [ ] Make sure all framework code uses `services.TryAddXyz(...)` and `services.TryAddEnumerableXyz(...)` to avoid multiple registrations of the same service
- [ ] Have the user's `MauiProgram` class call `.Build()` on the `MauiAppBuilder` and return the constructed `MauiApp`. This makes it clearer to the user that they fully baked the app themselves and the framework isn't going to do magic on it later. This is also more similar to ASP.NET Core.
- [ ] Look into using [hosted services](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostedservice?view=dotnet-plat-ext-5.0). This would replace the icky `IMauiInitializeService` that I added, and also to enable users to use hosted services. MAUI will automatically call StartAsync/StopAsync at the appropriate time. Sadly, this pattern is async-only, whereas in MAUI the call sites are always (or at least almost always) sync. But give that this is app startup/shutdown, doing async-over-sync might be OK (we would block).
- [ ] Add collections for Handlers, Fonts, and other collection-y things on `MauiAppBuilder` instead of the various `ConfigureXyz` methods. This is much simpler code for the user and avoids syntax goop that lambdas have.

mattleibow commented:
Great things in that last comment!

For the name we should at least have `MauiAndroidApplication`. 

One thing that needs to be handled is the case on Android there are 2 parts: `MauiApplication` and `MauiAppCompatActivity``. And I think iOS/MacCatalyst will have the same with the `UISceneDelegate` when we get multi-window. It is not possible to have an Android app without both the `Applictation` (which runs everything up to `Build()`) and then `Activity`(which starts the UI process).

Eilon committed:
Have user code return the built MauiApp

Instead of returning the builder.

Eilon committed:
Remove old comments

PureWeen commented:
@mattleibow can we go with

MauiNativeApplication => AppDelegate, Android.App.Application, Application
MauiNativeWindow => SceneDelegate, Activity, Window

for everything?  Possibly even drop the Maui prefix?

mhutch commented:
This is great. I love the more straightforward imperative feel.

This gave me an idea for using top level statements. If the body of `CreateMauiApp` became top level statements, a source generator could generate the `CreateMauiApp` method from them. My initial thought was that this would be far too magical, but we're already using source generators to obscure the real entrypoints, and I don't think it's any worse than that.

mattleibow commented:
> @mattleibow can we go with
> 
> MauiNativeApplication => AppDelegate, Android.App.Application, Application
> MauiNativeWindow => SceneDelegate, Activity, Window
> 
> for everything? Possibly even drop the Maui prefix?

YES! YES! and maybe!

We may want the Maui prefix to avoid clashes with other things the libraries do "Native" is very generic. But this is just my obsession with prefixes :)

NativeApplication is actually fine. I just had to write out that my deep-rooted, but still light (I hope), OCD wants things.

rmarinho commented:
Not a fan of. the "Native" in the namings.. I think the most used api should be MauiApplication and MauiWindow in all platforms

Redth commented:
> This is great. I love the more straightforward imperative feel.
> 
> This gave me an idea for using top level statements. If the body of `CreateMauiApp` became top level statements, a source generator could generate the `CreateMauiApp` method from them. My initial thought was that this would be far too magical, but AFAICT we're already using source generators to obscure the real platform-specific entrypoints, and I don't think it's any worse than that.

@mhutch Yes, this would be awesome!

The idea was always to source gen the Activity/AppDelegate/etc parts eventually...

Last I looked at the top level statements, I don't think that source gen has access to know its namespace/type/method to be able to call it?  I had thought the same once this PR was started - it would be really nice to have a top level statements file be called from the Activity/AppDelegate/etc entry points...

mhutch commented:
> Last I looked at the top level statements, I don't think that source gen has access to know its namespace/type/method to be able to call it? I had thought the same once this PR was started - it would be really nice to have a top level statements file be called from the Activity/AppDelegate/etc entry points...

IIRC the generated main method cannot be called from C#, but I assumed (maybe incorrectly) that source generators would be inspect the syntax tree of top level statements like they can other code. If a source generator can inspect the top level statements, it can generate a new method with those same statements (and appropriate `#line` directives), and the linker can remove the unused entrypoint method. If that doesn't work, invoking the entrypoint method via reflection is an option - not an ideal one, but it's a single call so probably not a big deal.

Another aspect to consider is how we would return the user-created app object from the top level statements so the generated `CreateMauiApp`  method can return it to the caller.  It also needs to make sense - even if we *could* do `return app` as a top level statement, that would be kind nonsensical as there's no context as to where the app is returned *to* and what happens next.

One way would be to generate a "return method" - a method the user must call on the app as the last toplevel statement. This could look like a mainloop, such as an `app.Run()` extension method, which would be cute but rather misleading about what's happening. Or we could be more explicit about what's happening e.g. `Platform.HostApp(app)`. If we used the approach of generating a `CreateMauiApp` method from the top level statements, we could directly replace the return method method call with a return statement. If we take the other approach and have the generated `CreateMauiApp` method somehow invoke the actual entrypoint method, then the return method could stash the app on a private static field from where the `CreateMauiApp` could retrieve it and return it.

Eilon commented:
@mhutch / @Redth - yes I'd love to see top-level statements and other simplifications in .NET MAUI! I logged https://github.com/dotnet/maui/issues/2243 to track this (I couldn't find an existing issue) because it's certainly out of scope of this PR, and I don't want to lose track of it.

Eilon committed:
Use proper DI APIs for registering/acquiring services

mattleibow reviewed (approved):
Code is looking good. Just waiting on the templates.

Eilon committed:
Update project templates to use new pattern

Eilon committed:
Fix build

mattleibow committed:
Merge remote-tracking branch 'origin/main' into eilon/startup-unification-rc1

# Conflicts:
#	src/Compatibility/Core/src/AppHostBuilderExtensions.cs

mattleibow committed:
Merge remote-tracking branch 'origin/main' into eilon/startup-unification-rc1

mattleibow committed:
Make sure we have a copy of the envvars

Works around https://github.com/xamarin/xamarin-macios/issues/12555
